proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+                                         (canDonate \<and> scPtrOptCallee = None
+                                                    \<and> scPtrOptDonated \<noteq> None \<longrightarrow>
+                                            calleePtr \<noteq> ksIdleThread s \<and>
+                                            the scPtrOptDonated \<noteq> idle_sc_ptr \<and>
+                                            (\<forall>ko. ko_at' ko (the scPtrOptDonated) s \<longrightarrow>
+                                                    scTCB ko \<noteq> Some (ksIdleThread s)))"
+                             in hoare_strengthen_post[where a="setThreadState _ _", rotated], force)
+                apply (wpsimp wp: gts_wp' threadGet_wp hoare_vcg_all_lift hoare_vcg_imp_lift'
+                            simp: setReplyTCB_def getReplyTCB_def)+
+  apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_tcb sym_refs_asrt_def)
+  apply (drule_tac ko=obj in sym_refs_ko_atD'[rotated])
+   apply (auto simp: valid_reply'_def valid_idle'_def idle_tcb'_def refs_of_rev'
+                     pred_tcb_at'_def ko_wp_at'_def obj_at'_def projectKO_eq projectKO_tcb)
+  done
+
+lemma replyPush_valid_queues:
+  "\<lbrace>valid_queues and valid_objs'\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_. valid_queues\<rbrace>"
+  by (rule monadic_rewrite_refine_valid[OF monadic_rewrite_replyPush, where P''=\<top>, simplified])
+     (wpsimp wp: weak_if_wp schedContextDonate_valid_queues hoare_drop_imps)
+
+lemma replyPush_valid_queues'[wp]:
+  "replyPush callerPtr calleePtr replyPtr canDonate \<lbrace>valid_queues'\<rbrace>"
+  by (wpsimp wp: schedContextDonate_valid_queues' hoare_drop_imp hoare_vcg_all_lift weak_if_wp
+           simp: replyPush_def)
+
+lemma replyPush_valid_release_queue[wp]:
+  "replyPush callerPtr calleePtr replyPtr canDonate \<lbrace>valid_release_queue\<rbrace>"
+  by (wpsimp wp: schedContextDonate_valid_release_queue hoare_drop_imp hoare_vcg_all_lift weak_if_wp
+           simp: replyPush_def)
+
+lemma replyPush_untyped_ranges_zero'[wp]:
+  "replyPush callerPtr calleePtr replyPtr canDonate \<lbrace>untyped_ranges_zero'\<rbrace>"
+  apply (clarsimp simp: untyped_ranges_zero_inv_null_filter_cteCaps_of)
+  apply (rule hoare_lift_Pf[where f="ctes_of"])
+   apply wp+
+  done
+
+lemma replyPush_invs':
+  "\<lbrace>\<lambda>s. invs' s \<and> sch_act_not callerPtr s \<and> st_tcb_at' active' callerPtr s \<and>
+        ex_nonz_cap_to' callerPtr s \<and> ex_nonz_cap_to' calleePtr s \<and> ex_nonz_cap_to' replyPtr s\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_ s. invs' s\<rbrace>"
+  unfolding invs'_def valid_state'_def valid_pspace'_def
+  by (wpsimp wp: replyPush_valid_objs' replyPush_sch_act_wf replyPush_valid_queues
+                 replyPush_if_live_then_nonz_cap' replyPush_valid_idle'
+           simp: valid_pspace'_def)
+
+lemma setThreadState_Running_invs':
```

Comment:
Can you generalise this to the case where the thread state is restart, or any other thread state? Is it because it's runnable? I'm not sure...

Reply:
This does seem interesting. I would like to have a closer look at this...